### PR TITLE
Remove table layouts from Hive

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java [hunk 1]
```diff
@@ -29,28 +34,76 @@
 {
     private final String schemaName;
     private final String tableName;
-
+    private final List<ColumnHandle> partitionColumns;
+    private final Optional<List<HivePartition>> partitions;
+    private final TupleDomain<HiveColumnHandle> compactEffectivePredicate;
```
Comment (takezoe) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: In my understanding, `TupleDomain` need to be JSON serializable if we add it to `TableHandle`, but currently some classes which can be included in `TupleDomain` aren't JSON seraializable. Actually I'm facing this problem in #511. Making all classes which can be included in `TupleDomain` JSON serializable is a right way?

Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: `TupleDomain` should work for all types. `Domain` contains `ValueSet` and the two implementations with values `EquatableValueSet` and `SortedRangeSet` handle this by storing the value as a `Block` (which is always serializable in the engine via `BlockJsonSerde`).

Comment (takezoe) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveTableHandle.java]: Oh, I see. Let me try to use `BlockJsonSerde`. Thanks!

File: presto-hive/pom.xml [hunk 2]
```diff
@@ -191,16 +191,6 @@
             <artifactId>javax.inject</artifactId>
         </dependency>
 
-        <dependency>
-            <groupId>com.fasterxml.jackson.core</groupId>
-            <artifactId>jackson-core</artifactId>
-        </dependency>
```
Comment (kokosing) [inline presto-hive/pom.xml]: How compile scope dependencies are related to test removal?

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -441,21 +441,41 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
     /**
      * @return whether delete without table scan is supported
      */
+    @Deprecated
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: Please add a comment that tells what should be used instead.

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 4]
```diff
@@ -441,21 +441,41 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
...
+    {
+        throw new PrestoException(NOT_SUPPORTED, "This connector does not support deletes");
+    }
+
     /**
      * Delete the provided table layout
      *
      * @return number of rows deleted, or null for unknown
      */
+    @Deprecated
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: same here

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 5]
```diff
@@ -120,11 +120,24 @@ default ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorT
      * the original partitioning handle associated with the provided table layout handle,
      * as promised by {@link #getCommonPartitioningHandle}.
      */
+    @Deprecated
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: same here

File: presto-hive/pom.xml (hunk 2 above)
Comment (electrum) [reply presto-hive/pom.xml]: The test was the only remaining usage. The dependency analyzer doesn't complain if you have a compile dependency that is only used by test code.

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddExchanges.java [hunk 6]
```diff
@@ -551,6 +552,16 @@ public PlanWithProperties visitValues(ValuesNode node, PreferredProperties prefe
                             .build());
         }
 
+        @Override
+        public PlanWithProperties visitMetadataDelete(MetadataDeleteNode node, PreferredProperties context)
+        {
+            return new PlanWithProperties(
+                    node,
+                    ActualProperties.builder()
+                            .global(singleStreamPartition())
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddExchanges.java]: These are needed because the node is now constructed earlier in the plan, before `AddExchanges` runs.

File: presto-main/src/main/java/io/prestosql/operator/MetadataDeleteOperator.java [hunk 7]
```diff
@@ -40,26 +38,23 @@
     {
         private final int operatorId;
         private final PlanNodeId planNodeId;
-        private final Metadata metadata;
-        private final Session session;
-        private final TableHandle tableHandle;
+        private final Supplier<OptionalLong> deleteAction;
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/operator/MetadataDeleteOperator.java]: This is gone now

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/MetadataDeleteNode.java [hunk 8]
```diff
@@ -16,40 +16,53 @@
...
 
 import javax.annotation.concurrent.Immutable;
 
 import java.util.List;
 
+import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 @Immutable
 public class MetadataDeleteNode
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/sql/planner/plan/MetadataDeleteNode.java]: Done

**Comment by martint** [comment]
This should wait for https://github.com/prestosql/presto/pull/731. The legacy PushPredicateIntoTableScan logic processes the predicates through DomainTranslator, which normalizes expressions in the same way UnwrapCastsInComparison in that PR does. However, the normalization is not done for the new `applyFilter`. Leaving casts "wrapped" causes some of the cost estimations to fail, so it affects the ability for the join ordering optimizer to do its job. #731 introduces the normalization for expressions anywhere in the plan, so it compensates for the "missing" behavior in handling applyFilter (arguably, it should have never been there in the first place)

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -1742,35 +1736,32 @@ private TableMetadata getTableMetadata(String catalog, String schema, String tab
...
-                    ConnectorTableLayoutHandle connectorLayout = metadata.getLayout(transactionSession, tableHandle.get(), Constraint.alwaysTrue(), Optional.empty())
-                            .get()
-                            .getNewTableHandle()
-                            .getLayout()
-                            .get();
-
-                    return propertyGetter.apply((HiveTableLayoutHandle) connectorLayout);
+                    QualifiedObjectName name = new QualifiedObjectName(catalog, TPCH_SCHEMA, tableName);
+                    TableHandle table = metadata.getTableHandle(transactionSession, name)
+                            .orElseThrow(AssertionError::new);
```
Comment (martint) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Provide a meaningful message with the exception?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -1742,35 +1736,32 @@ private TableMetadata getTableMetadata(String catalog, String schema, String tab
...
+                    table = metadata.applyFilter(transactionSession, table, Constraint.alwaysTrue())
+                            .orElseThrow(AssertionError::new)
```
Comment (martint) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Same here

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestFullParquetReader.java [hunk 11]
```diff
@@ -21,6 +21,6 @@
 {
     public TestFullParquetReader()
     {
-        super(ParquetTester.fullParquetTester());
+        super(new ParquetTester());
```
Comment (martint) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestFullParquetReader.java]: Unrelated change?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestParquetReader.java [hunk 12]
```diff
@@ -21,6 +21,6 @@
 {
     public TestParquetReader()
     {
-        super(ParquetTester.quickParquetTester());
+        super(new ParquetTester());
```
Comment (martint) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestParquetReader.java]: Unrelated change?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 13]
```diff
@@ -329,7 +332,11 @@ public ConnectorTableHandle getTableHandleForStatisticsCollection(ConnectorSessi
             }
         });
 
-        return handle;
+        HiveTableHandle table = handle;
+        return partitionValuesList
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Why is this needed now?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 14]
```diff
@@ -1653,29 +1616,52 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+    {
+        HiveTableHandle handle = (HiveTableHandle) tableHandle;
+        checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
+
+        if (handle.getPartitions().isPresent()) {
+            return Optional.empty(); // TODO: optimize multiple calls to applyFilter
+        }
+
+        HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, handle, constraint);
+        ConstraintApplicationResult<ConnectorTableHandle> result = new ConstraintApplicationResult<>(
```
Comment (martint) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Assign to `newHandle` directly and move this to the end of the method. So:

```java
HiveTableHandle newHandle = partitionManager.applyPartitionResult(handle, partitionResult);

if (handle.getPartitions().equals(newHandle.getPartitions()) &&
        handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate()) &&
        handle.getBucketFilter().equals(newHandle.getBucketFilter())) {
    return Optional.empty();
}

return Optional.of(new ConstraintApplicationResult<>(newHandle, partitionResult.getUnenforcedConstraint()));
```

Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Good idea, this is better

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestParquetReader.java (hunk 12 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestParquetReader.java]: Fixed

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestFullParquetReader.java (hunk 11 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestFullParquetReader.java]: Fixed

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 9 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Done

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 10 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Done

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java [hunk 15]
```diff
@@ -144,7 +141,7 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
                     ImmutableList.of(new HivePartition(tableName)),
                     compactEffectivePredicate,
                     effectivePredicate,
-                    none(),
+                    all(),
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java]: I'm not sure. I think this change is a mistake.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 13 above)
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: The listing has to be done somewhere. Doing it here makes sense as we'd otherwise need to move this into `HivePartitionManager.getOrLoadPartitions()`, with the special case logic for the analyze partition values list.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java [hunk 16]
```diff
@@ -126,15 +122,16 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
         TupleDomain<ColumnHandle> effectivePredicate = constraint.getSummary();
 
         SchemaTableName tableName = hiveTableHandle.getSchemaTableName();
-        Table table = getTable(metastore, tableName);
-        Optional<HiveBucketHandle> hiveBucketHandle = getHiveBucketHandle(table);
-
-        List<HiveColumnHandle> partitionColumns = getPartitionKeyColumnHandles(table);
+        Optional<HiveBucketHandle> hiveBucketHandle = hiveTableHandle.getBucketHandle();
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartitionManager.java]: You're right, all of these changes can be done separately. Done

**Comment by electrum** [comment]
@martint updated
